mrc-4747: Support parsing of complex queries

plietar opened the pull request:
Using the new python bindings for the Rust query parser, this adds
support for evaluating complex queries.

plietar committed:
mrc-4747: Support parsing of complex queries.

Using the new python bindings for the Rust query parser, this adds
support for evaluating complex queries.

plietar committed:
Update outpack schema.

plietar commented on pyproject.toml:
```diff
@@ -28,8 +28,11 @@ dependencies = [
   "jsonschema",
-  "pygit2"
+  "pygit2",
+  "outpack_query_parser @ git+https://github.com/mrc-ide/outpack_server@python-bindings"
```
The git url is temporary, until the python bindings get merged and we make a pip release for them.

plietar committed:
More tests and minor fix.

r-ash commented on src/outpack/search.py:
```diff
@@ -1,60 +1,244 @@
+def search(
+    query: Union[Query, str],
+    *,
+    root: OutpackRoot,
```
I think it would be useful to keep this as also being able to take a string, so end users can call it directly if they want to search for anything. Use case being someone has produced a bunch of packets and wants to find on disk the plot produced the last time they ran a report with `iso = "MWI"` so that they can share it with someone. 

Could also be worth having a wrapper around this search maybe which doesn't expose the `this` as might be confusing for that to be in the interface, and same for `environment` if/when it comes

plietar replied:
Good point. I hadn't thought of it being a user-facing API. Test coverage had flagged that codepath in `as_root` as unused so I'd removed it. I'll add it back with a test for it.

I noticed `as_root(..)` and `root_open(.., locate=False)` basically did the same thing, except the latter has more error checking so I've switched to that.

plietar replied:
I'm not really convinced exposing `this` is that harmful to users. It's optional anyway.

r-ash reviewed (commented):
This looks great, small comment about keeping a public `search` interface but this looks really solid

plietar committed:
Allow search to be called with a an str or Path root.

While internal calls generally already have an OutpackRoot, search can
also be called by users directly. Having the API also accept either an
str or a Path makes this more convenient.

root_open's locate parameter is turned into a named argument instead of
positional, and gains a default, making call sites clearer.

plietar committed:
Switch to published outpack_query_parser.

richfitz commented on pyproject.toml:
```diff
@@ -29,10 +29,8 @@ dependencies = [
   "jsonschema",
   "pygit2",
-  "outpack_query_parser @ git+https://github.com/mrc-ide/outpack_server@python-bindings"
+  "outpack-query-parser"
```
Can you leave a note somewhere (in the README and/or here) about how we use the git version so if we're developing something that needs it we can easily use it?

richfitz reviewed (approved):
Thanks - this is really great. Just a note here for me for the future. Thanks for all the tidying of the code too!

plietar committed:
Add note to README.
